### PR TITLE
[TypeReconstruction] Fix reconstruction for associated types.

### PR DESCRIPTION
**Pull request by dcci**
<rdar://problem/38931073>

### CODE REVIEW

**Comment by dcci** [comment]
@swift-ci test and merge

File: lib/IDE/TypeReconstruction.cpp [hunk 1]
```diff
@@ -748,6 +748,27 @@ static void VisitNodeAssociatedTypeRef(
...
+          if (!atd)
+            continue;
+          if (assoc_id == atd->getName()) {
+            associated = atd;
+            break;
+          }
+        }
+        if (associated) {
+          result._types.push_back(
+          DependentMemberType::get(protocol_type, assoc_id));
```
Comment (slavapestov) [inline lib/IDE/TypeReconstruction.cpp]: Indentation is wrong here. Also you can’t have a DependentMemberType with an existential base. When does such a mangling come up?

Comment (dcci) [reply lib/IDE/TypeReconstruction.cpp]: It shows up as type in DWARF (on some project I don't have access to anymore). You think we should just reject this outright?

Comment (slavapestov) [reply lib/IDE/TypeReconstruction.cpp]: Unless you can come up with a test case, in which case we need to see why it’s mangling such a type.

Comment (dcci) [reply lib/IDE/TypeReconstruction.cpp]: Makes sense, I'll send a new pull request soon. Thanks for your feedback!

**Pull request merged**